packages/plugin/vite/src/util/package.ts

```diff
@@ -0,0 +1,116 @@
+import fs from 'node:fs';
+import path from 'node:path';
+
+import type { PackageJsonManifest } from './packageJson';
+
+export interface Dependency {
+  name: string;
+  path: SourceAndDestination;
+  dependencies: Dependency[];
+}
+
+export interface SourceAndDestination {
+  src: string;
+  dest: string;
+}
+
+function isRootPath(dir: string) {
+  // Linux or Windows root path
+  return dir === '/' || /^[A-Z]:\\$/i.test(dir);
+}
+
+export async function lookupNodeModulesPaths(root: string, paths: string[] = []): Promise<string[]> {
+  if (!root) return paths;
+  if (!path.isAbsolute(root)) return paths;
+
+  const p = path.join(root, 'node_modules');
+
+  if (fs.existsSync(p) && (await fs.promises.stat(p)).isDirectory()) {
+    paths = paths.concat(p);
+  }
+  root = path.join(root, '..');
+
+  return isRootPath(root) ? paths : await lookupNodeModulesPaths(root, paths);
+}
+
+export async function readPackageJson(root = process.cwd()): Promise<PackageJsonManifest> {
+  const packageJsonPath = path.join(root, 'package.json');
+  try {
+    const packageJsonStr = await fs.promises.readFile(packageJsonPath, 'utf8');
+    try {
+      return JSON.parse(packageJsonStr);
+    } catch (error) {
+      console.error(`parse 'package.json': ${packageJsonPath}`);
+      throw error;
+    }
+  } catch (error) {
+    console.error(`'package.json' not found: ${packageJsonPath}`);
+    throw error;
+  }
```

Comment:
```suggestion
  try {
    const packageJsonStr = await fs.promises.readFile(packageJsonPath, 'utf8');
    return JSON.parse(packageJsonStr);
  } catch (error) {
    console.error(`parse ${packageJsonPath} failed: ${error}`);
    throw error;
  }
```

Don't worry, if the file doesn't exist, the error will be described in nodejs, so we don't need to add any additional error description.